### DISCUSSION TITLE
But what about performance? 🐌  - 30 min for about 3k SSG-pages normal? 🤔

### DISCUSSION THREAD

**Discussion by nickluger** [vercel/next.js#14122] [Help]
Via this thread I'd like to find out about and compare **SSG-build and deployment times for Next and Vercel.**

We render and deploy about **3k**  SSG-pages via **Vercel**. This takes about **30 - 35 
minutes**. Pages will grow to 15k this year I think, so I’m a bit concerned.

Build time (`Automatically optimizing pages...` →  `Done in xxx s`-log) is about **8 minutes** which can be improved of course depending on how we call our API in`getStaticProps` and how we cache stuff during the build.

`Uploading build outputs.` takes about **12-14 minutes** alone though and sometimes throws timeout errors on the AWS S3 side Vercel seems to be uploading to.

The remaining bout 10 minutes are filled with the initial setup stuff like `yarn install` and the logs at the end. `Build completed. Populating build cache...`, etc.

Vercel support told me this build time is unusually large and the largest Vercel-clients with much more SSG-pages would need only a **fraction of the time**.  (Edit: When the support wrote us our build time was about 50 minutes, still "a fraction" should be much lower, especially when taking a look who the "largest clients" are).

I would like to find out why our build is so slow and what makes yours faster.

So if you have a decent number of SSG pages: What's your performance and build time?

**Comment by c0b41** [comment]
checkout #11552

**Comment by Southclaws** [comment]
We have a similar situation, out site is approaching 2000 pages - it's predominately documentation but it has a lot of translations (and more to come!) stored in markdown and rendered with next-mdx-remote.

We originally moved away from Docusaurus and built a Next.js app because Docusaurus build times were 20+ minutes and Vercel was timing out. Our new site is around 4 minutes but some new additions recently pushed that towards the 10 minute mark.

Are you paying for Vercel's professional plan? We're using the free plan as it's a non-profit project with zero income/donations but I'm considering trying the Pro plan to see if build times would be improved.

**Reply by leerob** [reply]
Have you checked out Incremental Static Regeneration? That helps keep your build times low, while still keeping all the benefits of static!

**Reply by Southclaws** [reply]
Ah, I've heard of this and haven't read about it yet. Thanks I'll check it out!

**Reply by Southclaws** [reply]
Update: couldn't really improve it any further so we traded build time for request time and removed static pages and simply served markdown from an API and rendered it in getServerSideProps.

**Reply by ChenReuven** [reply]
So ssg is Not scale well if you move the SSR approach, right?

**Reply by leerob** [reply]
SSG can scale fine, it depends on your use case: https://www.smashingmagazine.com/2021/04/incremental-static-regeneration-nextjs/

**Reply by ChenReuven** [reply]
Hey :-)

I try the commerce Demo with local provider with 10000 generated local  products(regular loop over 10000 mock products) with ISR. It take more than 50 mintus approx just for the build. (I put the revalidate to 5 min) 

This is VERY slow :-(

What I can do beside that to improve the build?

Thanks

- your article is awesome!!!

**Reply by leerob** [reply]
You're generating all of those products at build time! Instead, set the getStaticPaths to `[]` and you won't build any of those products at build time. Then, use `fallback` to generate them on demand.

https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration

**Reply by amityweb** [reply]
But then the website is slow for the first people after the initial build. It would be great to have all pages fast after the build, and not require website visitors/customers to do wait for the generation.

**Comment by sanderfish** [comment]
I'm running into the same issue and received a similar response from support. We currently SSG ~2000 pages which takes 30-35 minutes. When I add another 1000 pages for a new section of the site we're building, Vercel times out at 45 minutes. A 45 minute build is undesirable in the first place.

One potential performance improve I see in our setup is the fetching of global data. We fetch global data (header, footer, site config) from a CMS on each page. We could fetch this pre-build, write it to a file and read from that file during SSG, but it should be fetched from the CMS during Incremental Static Regeneration. I'm not sure if it's possible to identify if getStaticProps is revalidating or SSG'ing.

While Incremental Static Regeneration is amazing, I would say it's desirable to statically generate as many pages as possible at build time. I understand it's a trade-off though.

**Reply by nickluger** [reply]
We use a workaround now and build and cache >6k pages in < 15 minutes. We have a GraphQL backend, which allows us to fetch all data for about 2-3 k pages in 1 API call. Something like a `getAllPosts`-query instead of `getPostBySlug` for example. Then we cache every data piece in Redis. This is all done before calling `next build`. `getStaticPaths` and `getStaticProps` only ever access Redis and never the API directly. This makes the build extremly fast and efficient at the same time.

**Reply by nickluger** [reply]
Yes you can detect whether you're running SSG build by inspecting the `CI` environment variable. This is set during the SSG build, but should not be set during execution of the page as a serverless function, as documented here:
https://vercel.com/docs/environment-variables#system-environment-variables

**Reply by sanderfish** [reply]
Thanks for your reply! That sounds similar to what I'm describing as a potential performance improvement we could make. How do you make sure the Redis cache is up to date? Is it updated only during build? What are you doing to have it serve up to date content during Incremental Static Regeneration?

**Reply by nickluger** [reply]
We have no requirement or possibility for iSSG currently, because we have rather largish and infrequent publish steps on our app's content which require multiple pages to be rebuilt at the same time, because they reference each other. This is not possible with iSSG currently anyway, and if we used it our pages could become out-of-sync.

So the Single Source of Truth for data in our case is always the latest full rebuild, and we trigger one on every content publish. Other, smaller, more current content updates or user specific content we simply fetch client-side, which makes it up-to-date.

Also, we deploy code multiple times a day which makes iSSG obsolete anyway, because all pages are cleared from cache on a code change rebuild. So we use Redis just as a tool to speed up the build, not as a runtime cache.

But of course your API could keep the cache up to date on changes, by writing to it. (Same as it would, if you would use Redis as normal application cache, as most people use it 😄)

**Reply by sanderfish** [reply]
That makes sense. By using the `CI` environment variable we should be able to fetch accordingly, either from cache of fresh from the API/CMS.

Thanks Nick!

**Reply by nickluger** [reply]
Glad, if I could help. Working around the limitations of the (otherwise superb) Jamstack was sometimes tedious for me, too. 😅

If you only use the cache during the SSG build step, you could also work with local files I assume. This should be faster and easier than a Redis cache. Just don't put em into the `public`-directory, as they would be deployed with every serverless function (i.e. Page) which  would slow them down and they have a 250 MB limit, too.

**Comment by GeeWee** [comment]
We have the same issue with 12k~ pages. We can build them in around 15 minutes with some clever caching, but the step 
`Deploying build outputs...` takes more than 15 minutes and then the deployment fails.

**Comment by francoisruty** [comment]
I'm encountering same problem, large-scale site (dozens of thousands of pages, soon hundreds of thousands) but I want to keep everything under SSG
- what I did already: I designed my own incremental SSG system, I track what has changed in a DB and then my API exposes only the pages that have changed via getStaticPaths, which means every build will only work on pages that have changes
- what I'm considering doing: problem is that I have also pages with content that changes frequently. For those I still want to do SSG, but I'm considering not rebuilding the html, just changing the content
I noticed html pages after SSG contain a js script with id __NEXT_DATA__ that contain all the data so I'm wondering whether I could just edit this directly so that I can update the content without rebuilding the html itself

**Reply by hahazexia** [reply]
have you found better way to reduce the SSG build time?
i want to keep everything under SSG too, i don't want to use something like ISR，because ISR will need nodejs backend on server side.
